ONSdigital/eq-questionnaire-runner
# Using supplementary data list items

### What is the context of this PR?
This PR implements the usage of supplementary data list items in a survey. This is an extension to the [non list items pr 
](https://github.com/ONSdigital/eq-questionnaire-runner/pull/1169)
The main functional changes in this ticket, are:

- altering the error handling in `supplementary_data_store.py` so that nested data can be optional, such as the `guidance_exclude` block, but still raising an error if you try to index non-dictionary non-none data with a selector. 
- Adding a method to check if the data being referenced is repeating.
- Updating `value_source_resolver.py` to make sure that a list item id is used if the data being referenced is repeating.

Other than that, the test data has been combined to reduce the number of files, and unit tests and functional tests have been added.

### How to review
Read through the trello card and check that the test schema allows for testing the mentioned requirements
Check the changes to value source resolver and supplementary data store and that the changes are suitably tested

### How to test functionally
Exactly as with the other PR: following the guidance in run-mock-sds-endpoint.md start the mock sds api and pick the `test_supplementary_data` schema. 
The dropdown for sds dataset id will show two different ids, these are the same structure but with different data, check that placeholders render correctly with each of them. You should also be able to preserve your response id, relaunch the survey with the updated dataset which has some different data and see your original answers against the new placeholders. In this instance, check that the list items have updated correctly and any dynamic answers or repeating sections are updated/removed/added accordingly

Test any unhappy path you can think of, and make sure the functional test passes locally (removing the skip)

### Checklist
* [ ] The new functional test passes if you uncomment the .skip and have the mock endpoint running
